// enum/index.js
var Enum = {};

var weapons = {
    cannon: "cannon",
    railgun: "railgun",
    mine: "mine"
}

Enum.weapons = weapons;

var weaponDamage = {};
weaponDamage[weapons.cannon] = 10;
weaponDamage[weapons.railgun] = 50;
weaponDamage[weapons.mine] = 100;
Enum.weaponDamage = weaponDamage;

Enum.startAmmo = {};
Enum.startAmmo.cannon = 100;
Enum.startAmmo.mine = 100;

Enum.playerStep = 8;

Enum.fieldWidth = Enum.fieldHeight = 512;

Enum.blockSize = 32;

Enum.lastUpdateInterval = 40;

Enum.bulletSpeed = 10;

Enum.directions = {
    up: "up",
    down: "down",
    left: "left",
    right: "right"
};

//Enum.bonuses = ["shield", "heart"];

Enum.bonuses = {
    shield: {
        title: "shield",
        value: 30
    },
    heart: {
        title: "heart",
        value: 20
    },
    crate: {
        title: "crate",
        value: 0
    }
}

Enum.bonusArray = [];

for (var key in Enum.bonuses) {
    Enum.bonusArray.push(Enum.bonuses[key].title);
}

Enum.crateBonuses = {
    cannon: {
        title: "cannon",
        value: 20,
        chance: 0
    },
    mine: {
        title: "mine",
        value: 2,
        chance: 25
    },
    invincibility: {
        title: "invincibility",
        value: 0,
        chance: 75
    }

}

Enum.playerHp = 100;

Enum.playerArmor = 0;

module.exports = Enum;
